redash/serializers/__init__.py

```diff
@@ -25,21 +25,22 @@ def public_widget(widget):
         'created_at': widget.created_at
     }
 
-    if widget.visualization and widget.visualization.id:
-        query_data = models.QueryResult.query.get(widget.visualization.query_rel.latest_query_data_id).to_dict()
+    v = widget.visualization
+    if v and v.id:
         res['visualization'] = {
-            'type': widget.visualization.type,
-            'name': widget.visualization.name,
-            'description': widget.visualization.description,
-            'options': json_loads(widget.visualization.options),
-            'updated_at': widget.visualization.updated_at,
-            'created_at': widget.visualization.created_at,
+            'type': v.type,
+            'name': v.name,
+            'description': v.description,
+            'options': json_loads(v.options),
+            'updated_at': v.updated_at,
+            'created_at': v.created_at,
             'query': {
+                'id': v.query_rel.id,
                 'query': ' ',  # workaround, as otherwise the query data won't be loaded.
```

Comment:
Still needed?

Reply:
Yes, otherwise dropdown won't work.

Reply:
Why?

Reply:
Ohhh, you were referring to the empty query text workaround? I thought you were referring to the query id. Anyway, the workaround can totally be removed - done in 16be2f6